File: StriverA2Z/05. Strings/1. Basic Easy/2. Palindrome Check.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <stack>
using namespace std;

class Solution {
public:
    string reverseWords(string s) {

        // remove leading and trailing spaces
        int start = s.find_first_not_of(' ');  // index of first nonspace char
        int end = s.find_last_not_of(' ');     // index of last nonspace char
        s = s.substr(start, end - start + 1);  // creates a substring from start to end index substr(pos, length)

        // split by whitespaces 
        stringstream ss(s);   // creates a stringstream taht allows to extract words seperated by spaces
        vector<string> words;
        string word;

        while (ss>>word){
            words.push_back(word);
        }

        // Reverse the words
        reverse(words.begin(), words.end());


        // Join with a single space
        string result; 
        for (int i = 0; i< words.size(); i++){
            result +=words[i];
            if (i+1 < words.size()) result += " ";  // add spcae only between words
        }

        return result;
    }
};

File: StriverA2Z/05. Strings/1. Basic Easy/2. Palindrome Check.py


class Solution:
    def reverseWords(self, s: str) -> str:

        s = s.strip()   # remove leading and trailing spaces and returns a new string
        words = s.split(" ")  # splits words based on space and returns a list of words
        # but this also includes empty strings in the list if there are multiple spaces between words
        
        new_s = " ".join(words[::-1]) # joins the words in reverse order with a single space in between
        
        new_s = new_s.strip() #not actually required d

        return new_s
        

class Solution:
    def reverseWords(self, s: str) -> str:
        words = s.strip().split() # splits words based on whitespace and removes extra spaces
        return " ".join(words[::-1]) # joins the words in reverse order with a single space in between